File: retos-programacion-2023-rendimiento/python/facil/reto_0.py
# Reto #0: EL FAMOSO "FIZZ BUZZ"
#### Dificultad: Fácil | Publicación: 26/12/22 | Corrección: 02/01/23

"""
Enunciado

 * Escribe un programa que muestre por consola (con un print) los
 * números de 1 a 100 (ambos incluidos y con un salto de línea entre
 * cada impresión), sustituyendo los siguientes:
 * - Múltiplos de 3 por la palabra "fizz".
 * - Múltiplos de 5 por la palabra "buzz".
 * - Múltiplos de 3 y de 5 a la vez por la palabra "fizzbuzz".

"""

import numpy as np
from timeit import default_timer

# solución pusheada, tipo = por comprension

t_inicial = default_timer()

for numero in range(1,101):
    if numero%3==0 and numero%5==0:
        print('fizzbuzz')
    elif numero%3==0:
        print('fizz')
    elif numero%5==0:
        print('buzz')
    else:
        print(numero)

t_final = default_timer()
t_resultado1 = t_final - t_inicial

# solución 2, tipo = funcional

def fizzbuzz():
    for numero in np.arange(1,101):
        if numero%3==0 and numero%5==0:
            print('fizzbuzz')
        elif numero%3==0:
            print('fizz')
        elif numero%5==0:
            print('buzz')
        else:
            print(numero)

t_inicial = default_timer()
fizzbuzz()
t_final = default_timer()
t_resultado2 = t_final - t_inicial

print("\nTiempo solucion pusheada: {} \nTiempo solucion 2: {}".format(t_resultado1,t_resultado2))

File: retos-programacion-2023-rendimiento/javascript/facil/reto_0.js
/*
* Escribe un programa que muestre por consola (con un print) los
* números de 1 a 100 (ambos incluidos y con un salto de línea entre
* cada impresión), sustituyendo los siguientes:
* - Múltiplos de 3 por la palabra "fizz".
* - Múltiplos de 5 por la palabra "buzz".
* - Múltiplos de 3 y de 5 a la vez por la palabra "fizzbuzz".
*/



const fizzbuzz = () => {
    for (let numero = 1; numero <= 100; numero++) {    
        if (numero % 3 == 0 && numero % 5 == 0) console.log('fizzbuzz')
        else if(numero % 3 == 0) console.log('fizz')            
        else if (numero % 5 == 0) console.log('buzz')            
        else console.log(numero)
    }
}

console.time()
fizzbuzz()
// console.timeLog()
console.timeEnd()

//solucion 2 por keys4dev

// [...Array(101).keys()] --> Crear un array de 0 a 100
// .slice(1) --> la corta del 1 en adelante para tener los numeros del 1 al 100
//.map permite ejecutar una operacion por valor en el array 
const deUnoACien = [...Array(101).keys()].slice(1).map(number =>{
    // Si el numero es multiplo de 3 y 5 guardo fizzbuzz en el array
    if ((number % 3 === 0) && (number % 5 === 0)){
        return number = `fizzbuzz`;
    // Sino si el numero es multiplo de 3 guardo fizz en el array
    }else if (number % 3 === 0){
        return number = `fizz`;
    // Sino si el numero es multiplo de 5 guardo buzz en el array
    }else if(number % 5 === 0){
        return number = `buzz`;
    }else{
        // Sino se da ninguna condicion paso el numero a string para imprimirlo despues
        return number.toString();
    }
});


console.time()
//Hago un map para hacer la operacion de imprimir por pantalla cada valor del array
deUnoACien.map(number => console.log(number));
console.timeEnd()

//Si quieres ver el contenido del array, descomenta la siguiente linea
console.log(deUnoACien); // el tipo de dato es un objeto